nsfc-plugin/src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java

```diff
@@ -0,0 +1,510 @@
+            inspectionPortPairGroupId = portPairGroup.getId();
+        } else {
+
+            if (!portPairGroup.getPortPairs().contains(portPair.getId())) {
+                portPairGroup.getPortPairs().add(portPair.getId());
+            }
+
+            this.osCalls.updatePortPairGroup(portPairGroup.getId(), portPairGroup);
+        }
+
+        NetworkElementImpl ingressElement = null;
+        NetworkElementImpl egressElement = null;
+
+        if (ingress != null) {
+            ingressElement = new NetworkElementImpl(ingress.getElementId(), ingress.getMacAddresses(),
+                                                     ingress.getPortIPs(), ingress.getParentId());
+        }
+
+        if (egress != null) {
+            egressElement = new NetworkElementImpl(egress.getElementId(), egress.getMacAddresses(),
+                    egress.getPortIPs(), egress.getParentId());
+        }
+
+        // Only parent id of the return value is ever used
+        PortPairGroupElement ppgElement = new PortPairGroupElement(inspectionPortPairGroupId);
+        PortPairElement retVal = new PortPairElement(portPair.getId(), ppgElement, ingressElement, egressElement);
+        ppgElement.getPortPairs().add(retVal);
+        return retVal;
+    }
+
+    @Override
+    public void removeInspectionPort(InspectionPortElement inspectionPort)
+            throws NetworkPortNotFoundException, Exception {
+        if (inspectionPort == null) {
+            LOG.warn("Attempt to remove a null Inspection Port");
+            return;
+        }
+
+        PortPair portPair = this.utils.fetchPortPairForInspectionPort(inspectionPort);
+
+        if (portPair != null) {
+            PortPairGroup portPairGroup = this.utils.fetchContainingPortPairGroup(portPair.getId());
+
+            if (portPairGroup != null) {
+                portPairGroup.getPortPairs().remove(portPair.getId());
+
+                if (portPairGroup.getPortPairs().size() > 0) {
+                    PortPairGroup ppgUpdate = Builders.portPairGroup().portPairs(portPairGroup.getPortPairs()).build();
+                    this.osCalls.updatePortPairGroup(portPairGroup.getId(), ppgUpdate);
+                } else {
+                    PortChain portChain = this.utils.fetchContainingPortChain(portPairGroup.getId());
+
+                    if (portChain != null) {
+                        List<String> ppgIds = portChain.getPortPairGroups();
+                        ppgIds.remove(portPairGroup.getId());
+
+                        // service function chain with with no port pair should be allowed to exist?
+                        PortChain portChainUpdate = Builders.portChain().portPairGroups(ppgIds).build();
+                        this.osCalls.updatePortChain(portChain.getId(), portChainUpdate);
+                    }
+                    this.osCalls.deletePortPairGroup(portPairGroup.getId());
+                }
+            }
+
+            this.osCalls.deletePortPair(portPair.getId());
```

Comment:
if we get a bad action response...where is the exception? Please fix all places.